kube-app-testing.sh

```diff
@@ -244,13 +251,190 @@ delete_kind_cluster () {
+  elif [[ "$1" == "nodepool" ]]; then
+  cat <<EOF
+{
+  "name": "${CLUSTER_NAME}",
+  "availability_zones": {
+    "number": 1
+  },
+  "scaling": {
+    "min": ${SCALING_MIN},
+    "max": ${SCALING_MAX}
+  }
+}
+EOF
+  # payload for creating a client key pair
+  elif [[ "$1" == "keypair" ]]; then
+  cat <<EOF
+{
+  "description": "CI-generated key pair",
+  "ttl_hours": 6,
+  "certificate_organizations": "system:masters"
+}
+EOF
+  # template a kubeconfig based on the created key pair.
+  # expects a file path as the second argument
+  elif [[ "$1" == "kubeconfig" ]]; then
+  cat > $2 << EOF
+apiVersion: v1
+clusters:
+- cluster:
+    certificate-authority-data: "${}"
+    server: "${}"
+  name: giantswarm-${CLUSTER_ID}
+contexts:
+- context:
+    cluster: giantswarm-${CLUSTER_ID}
+    user: giantswarm-${CLUSTER_ID}-user
+  name: giantswarm-${CLUSTER_ID}
+current-context: giantswarm-${CLUSTER_ID}
+kind: Config
+preferences: {}
+users:
+- name: giantswarm-${CLUSTER_ID}-user
+  user:
+    client-certificate-data: "${}"
+    client-key-data: "${}"
+EOF
+  fi
+}
+
+create_gs_cluster () {
+  # create a new cluster
+  curl ${GS_API_URI}/v5/clusters/ -X POST \
+    -H "Content-Type: application/json" \
+    -H "Authorization: giantswarm ${GSAPI_AUTH_TOKEN}" \
+    -d "$(gen_gs_data cluster)"
+
+  if [[ "$?" -gt 0 ]]; then
+    err "Cluster creation failed."
+    exit 3
+  fi
+
+  # get the list of clusters
+  _cluster_list=$(curl -s ${GS_API_URI}/v4/clusters/ -H "Authorization: giantswarm ${GSAPI_AUTH_TOKEN}")
+  # base64 encode the name to work around spaces
+  _base64_cluster_name=$(echo ${CLUSTER_NAME} | base64)
+
+  # we have to pass cluster dictionaries around as base64 encoded objects
+  # to avoid issues with whitespace characters.
+  for cluster in $(echo "${_cluster_list}" | jq -r '.[] | @base64'); do
+    unset _this_cluster
+
+    _jq() {
+      echo ${cluster} | base64 --decode | jq -r ${1}
+    }
+
+    _this_cluster=$(echo $(_jq '.name') | base64)
+
+    if [[ "$_base64_cluster_name" == "$_this_cluster" ]]; then
+      CLUSTER_ID=$(echo $(_jq '.id'))
+      break
+    fi
+  done
+
+  if [[ -z $CLUSTER_ID ]]; then
+    err "Cluster name matching '${CLUSTER_NAME}' could not be found."
+    exit 3
+  fi
+
+  # create a nodepool
+  curl ${GS_API_URI}/v5/clusters/${CLUSTER_ID}/nodepools/ -X POST \
+    -H "Content-Type: application/json" \
+    -H "Authorization: giantswarm ${GSAPI_AUTH_TOKEN}" \
+    -d "$(gen_gs_data nodepool)"
+
+  if [[ "$?" -gt 0 ]]; then
+    err "Nodepool creation failed."
+    exit 3
+  fi
+
+  # wait for the cluster to be ready
+  # declare a counter
+  _counter=0
+  until [ `curl -s -H "Authorization: giantswarm ${GSAPI_AUTH_TOKEN}" ${GS_API_URI}/v5/clusters/${CLUSTER_ID}/ | jq .conditions | grep -i "created" | wc -l` -gt 0 ]; do
+    # exit if the cluster hasn't been created in 30 minutes
+    if [[ "$_counter" -gt 60 ]]; then
+      err "Cluster not created after 30 minutes."
+      exit 3
```

Comment:
30 minutes was a relatively arbitrary time to wait - most clusters are up in 15-20 minutes.